src/main/java/org/dependencytrack/persistence/ComponentQueryManager.java

```diff
@@ -156,56 +159,178 @@ public PaginatedResult getComponents(final Project project, final boolean includ
+            }
+            if (orderBy.equalsIgnoreCase("name")) {
+                queryString +=
+                        """
+                            ORDER BY "NAME"
+                        """;
+            }
+            if (orderBy.equalsIgnoreCase("componentMetaInformation.integrityMatchStatus")) {
+                queryString +=
+                        """
+                            ORDER BY "INTEGRITY_CHECK_STATUS"
+                        """;
+            }
+            if (queryString.contains("ORDER BY")) {
+                if (orderDirection == OrderDirection.ASCENDING) {
+                    queryString += " ASC ";
+                } else if (orderDirection == OrderDirection.DESCENDING) {
+                    queryString += " DESC ";
+                }
+                queryString += """
+                        , "ID"
+                        """;
+            }
         }
-        if (includeMetrics) {
-            // Populate each Component object in the paginated result with transitive related
-            // data to minimize the number of round trips a client needs to make, process, and render.
-            for (Component component : result.getList(Component.class)) {
+
+        if (pagination != null && pagination.isPaginated()) {
+            queryString +=
+                    """
+                        OFFSET %d
+                        LIMIT %d;
+                    """.formatted(pagination.getOffset(), pagination.getLimit());
+        }
+
+        Query<?> query = pm.newQuery(Query.SQL, queryString);
+        query.setParameters(project.getId());
+        List<ComponentProjection> resultSet;
+        try {
+            resultSet = List.copyOf(query.executeResultList(ComponentProjection.class));
+        }
+        finally {
+            query.closeAll();
+        }
+        for (final var result : resultSet) {
+            final org.dependencytrack.model.Component component = ComponentUtil.mapToComponent(result);
+            if (includeMetrics) {
+//             Populate each Component object in the paginated result with transitive related
+//             data to minimize the number of round trips a client needs to make, process, and render.
                 component.setMetrics(getMostRecentDependencyMetrics(component));
                 final PackageURL purl = component.getPurl();
                 if (purl != null) {
                     final RepositoryType type = RepositoryType.resolve(purl);
                     if (RepositoryType.UNSUPPORTED != type) {
                         final RepositoryMetaComponent repoMetaComponent = getRepositoryMetaComponent(type, purl.getNamespace(), purl.getName());
                         component.setRepositoryMeta(repoMetaComponent);
-                        component.setComponentMetaInformation(getMetaInformation(component.getUuid()));
                     }
                 }
             }
+            componentsResult.add(component);
         }
```

Comment:
Can't we populate this from the SQL query's result as well?

Reply:
There are two separate tables 'DependencyMetrics' and 'Repository_meta_component' being queried in this. In order to populate these objects from same SQL query we can add more joins on these 2 tables.

Reply:
Not including DependencyMetrics query join, it'll complicate the query and make it less readable/maintainable.
Querying repo_meta_component is problematic as well because it'll need 3 fields (from parsing a purl) -> name, namespace and repository_type which is an enum. We wont have this enum value while building the query, it is calculated separately depending on component purl.